{{ cookiecutter.package_name|slugify }}/{% if cookiecutter.continuous_integration == "GitLab" %}.gitlab-ci.yml{% endif %}

```diff
@@ -25,37 +25,41 @@ variables:
-      if [ $DOCKER_PUSH -ne 1 ]; then
+      echo "$DOCKER_REGISTRY_PASSWORD" | docker login --username "$DOCKER_REGISTRY_USER" --password-stdin "$DOCKER_REGISTRY"
+      DOCKER_PUSH=${DOCKER_PUSH:-$(timeout 2s docker pull "$DOCKER_IMAGE":"$DOCKER_IMAGE_SHA" >/dev/null 2>&1 && echo $? || echo $?)}
+      if [ "$DOCKER_PUSH" -ne 1 ]; then
         echo "$DOCKER_IMAGE:$DOCKER_IMAGE_SHA exists, skipping this job..."
       else
         # Compile a list of image tags consisting of a hash of its contents, the latest tag if this
         # pipeline is running for the default branch, and the Git tag if this commit is tagged.
-        DOCKER_TAGS=( "$DOCKER_IMAGE_SHA" )
-        if [[ "$CI_COMMIT_BRANCH" == "$CI_DEFAULT_BRANCH" ]]; then DOCKER_TAGS+=( "latest" ); fi
-        if [[ -n $CI_COMMIT_TAG ]]; then DOCKER_TAGS+=( "$CI_COMMIT_TAG" ); fi
-        if [[ -n $CI_ENVIRONMENT_NAME ]]; then DOCKER_TAGS+=( "$CI_ENVIRONMENT_NAME" ); fi
+        DOCKER_TAGS="$DOCKER_IMAGE_SHA"
+        if [ "$CI_COMMIT_BRANCH" = "$CI_DEFAULT_BRANCH" ]; then DOCKER_TAGS="${DOCKER_TAGS} latest"; fi
+        if [ -n "$CI_COMMIT_TAG" ]; then DOCKER_TAGS="${DOCKER_TAGS} $CI_COMMIT_TAG"; fi
+        if [ -n "$CI_ENVIRONMENT_NAME" ]; then DOCKER_TAGS="$DOCKER_TAGS $CI_ENVIRONMENT_NAME"; fi
+        DOCKER_TAGS_JOINED=""
+        for DOCKER_TAG in ${DOCKER_TAGS}; do
+          DOCKER_TAGS_JOINED="${DOCKER_TAGS_JOINED} --tag $DOCKER_IMAGE:${DOCKER_TAG}"
+        done
+
         # Build the Docker image with all of the selected tags.
-        DOCKER_TAGS_JOINED=${DOCKER_TAGS[@]}
-        DOCKER_TAGS_JOINED="--tag $DOCKER_IMAGE:${DOCKER_TAGS_JOINED// / --tag $DOCKER_IMAGE:}"
         DOCKER_BUILDKIT=1 docker build \
           {%- if cookiecutter.with_fastapi_api|int or cookiecutter.with_streamlit_app|int %}
-          --build-arg APP_BASE_IMAGE=${DOCKER_BASE_IMAGE:-ci} \
+          --build-arg APP_BASE_IMAGE="${DOCKER_BASE_IMAGE:-ci}" \
           {%- endif %}
-          --build-arg SOURCE_BRANCH=$CI_COMMIT_REF_NAME \
-          --build-arg SOURCE_COMMIT=$CI_COMMIT_SHA \
-          --build-arg SOURCE_TIMESTAMP=$CI_COMMIT_TIMESTAMP \
+          --build-arg SOURCE_BRANCH="$CI_COMMIT_REF_NAME" \
+          --build-arg SOURCE_COMMIT="$CI_COMMIT_SHA" \
+          --build-arg SOURCE_TIMESTAMP="$CI_COMMIT_TIMESTAMP" \
           {%- if cookiecutter.private_package_repository_name %}
           --secret id=poetry-http-basic-username,env=POETRY_HTTP_BASIC_{{ cookiecutter.private_package_repository_name|upper }}_USERNAME \
           --secret id=poetry-http-basic-password,env=POETRY_HTTP_BASIC_{{ cookiecutter.private_package_repository_name|upper }}_PASSWORD \
           {%- endif %}
-          --target $DOCKER_TARGET \
+          --target "$DOCKER_TARGET" \
           --pull \
           $DOCKER_TAGS_JOINED \
           .
+
         # Push all the tagged images.
-        for DOCKER_TAG in ${DOCKER_TAGS[@]}; do
+        for DOCKER_TAG in ${DOCKER_TAGS}; do
```

Comment:
1. Could you remove the curly braces where they are not needed in this script (cf. example below)?
2. I saw you wrapped environment variables with `"`, which I think is a good idea -- would do that on this line too:
```suggestion
        for DOCKER_TAG in "$DOCKER_TAGS"; do
```

Reply:
Done.